### ISSUE TITLE
Loading already loaded "cardboard"/"daydream" VR device forcibly exits VR mode (SwitchToVR of the wiki fails if already it is on VR mode)

### ISSUE THREAD

**Issue by chelder86** [googlevr/gvr-unity-sdk#826]
### Found using:
 * Google VR SDK version: GoogleVRForUnity_1.110.0
 * Unity version: 2017.3.0f3 Personal
 * Phone manufacturer, model, and O/S version: Xiaomi 4A
 * Viewer manufacturer & model: Heromask

___
### Steps to reproduce the issue:

 1. Load Scene 1 and switch it on runtime to cardboard with the [wiki code](https://github.com/googlevr/gvr-unity-sdk/wiki/Switching-between-VR-and-non-VR-at-runtime/6339744e1eb4f3d1ecb0d8880cbf76ba27f75ad0)
 1. Load Scene 2 (also cardboard)
 1. Load Scene 1 and switch it on runtime to cardboard with the [wiki code](https://github.com/googlevr/gvr-unity-sdk/wiki/Switching-between-VR-and-non-VR-at-runtime/6339744e1eb4f3d1ecb0d8880cbf76ba27f75ad0). Notice it was already in cardboard mode.
 1. It will be loaded to "none" instead "cardboard"

___
### Workarounds:

Surround the code to check if it is already in cardboard mode. So:

    IEnumerator SwitchToVR() {
      if (UnityEngine.XR.XRSettings.loadedDeviceName != "cardboard")
        {
          XRSettings.LoadDeviceByName("cardboard"); // Or "daydream" (both lowercase).

          // Wait one frame!
          yield return null;

          // Now it's ok to enable VR mode.
          XRSettings.enabled = true;
        }
    }

**Comment by fredsa** [comment]
Can you include the code from Scene 1, but without the workaround?

If `UnityEngine.XR.XRSettings.loadedDeviceName == "cardboard"` is should still be okay to call ` XRSettings.LoadDeviceByName("cardboard")`. The **Cardboard** VR Device should remain loaded.

Note, when the **None** VR Device is loaded, you will see that `UnityEngine.XR.XRSettings.loadedDeviceName == ""` (i.e. the empty string `""`, and not `"none"` as you might expect).

**Comment by chelder86** [comment]
IEnumerator SwitchToVR()
    {


            UnityEngine.XR.XRSettings.LoadDeviceByName("cardboard"); // Or "daydream" (both lowercase).

            // Wait one frame!
            yield return null;

            // Now it's ok to enable VR mode.
            UnityEngine.XR.XRSettings.enabled = true;
        
	}


   Start(){ // Notice I also tried this in the Awake function
     // (...)
        StartCoroutine(SwitchToVR());
     // (...)    
   }

**Comment by chelder86** [comment]
The previous code (without surrounding it with the if clause), was working with the previous version of Google Cardboard and Unity, by the way.

**Comment by fredsa** [comment]
Thanks, I was able to reproduce this in Unity 2017.3.0p1

I filed Unity case 984886 to track this.

**Comment by fredsa** [comment]
I updated https://github.com/googlevr/gvr-unity-sdk/wiki/Switching-between-VR-and-non-VR-at-runtime to include a workaround for this issue.

**Comment by chelder86** [comment]
Happy to help :)

**Comment by numenavr** [comment]
Hi. Quick question: I'm using your SwitchToVR() coroutine but I'm getting the "Assertion failed: Assertion failed on expression: 'IsMatrixValid(matrix)'" error. Any ideas how to fix it? Thanks -

**Comment by fredsa** [comment]
@numenavr Please file a separate issue, and include repo steps and detailed log messages there.

**Comment by srivers8424** [comment]
If using Unity 2017.1.0f3 I'm guessing you would just replace all "XR" with "VR"?

**Comment by chelder86** [comment]
@srivers8424 with the previous version I was using (I used VR in that version), that bug did not occur, so I did not need to surround it with an if clause.

**Comment by srivers8424** [comment]
@chelder86  I was having the same issue, but wrapping it as you suggested worked like a charm. The mistake I made was checking for cardboard first instead of "None" or "" first. I included both because at the time I couldn't remember where I had used "None" or if it was obsolete.  Thank you for all your help! 

P.S. Here is the use of the [Wiki's ](https://github.com/googlevr/gvr-unity-sdk/issues/826)code I implemented in case anyone using Unity 2017.01f3 or other versions before 2017.3


  IEnumerator LoadDevice(string newDevice)
    {
        VRSettings.LoadDeviceByName(newDevice);
        yield return null;

        if (newDevice == "None" || newDevice == "")
        {
            VRSettings.enabled = false;
            Camera.main.ResetAspect();
        }
        else
        {
            VRSettings.enabled = true;
        }
    }

then call StartCoroutine(LoadDevice("cardboard")) or StartCoroutine(LoadDevice("") )

**Comment by fredsa** [comment]
Magic Window is now documented here:
https://developers.google.com/vr/develop/unity/guides/magic-window

**Comment by fredsa** [comment]
Leaving this issue open to track Unity case 984886, for the issue where calling `LoadDeviceName(XRSettings.loadedDeviceName)` causes that VR device to unexpectedly unload, rather than doing nothing.

**Comment by fredsa** [comment]
This issue has been fixed in Unity (verified in 2017.3.1p4). A warning message is now displayed in the console when attempting to load an already loaded VR device:

```
… W Unity : Invalid request to load VR Device daydream that is already loaded.
            This might cause the device to unload on some device configurations.
```

To enter VR mode at runtime see the sample code in the Unity reference docs ([XRSettings.LoadDeviceByName](https://docs.unity3d.com/ScriptReference/XR.XRSettings.LoadDeviceByName.html)), or for additional sample code, see the [Switching to VR mode at runtime](https://developers.google.com/vr/develop/unity/guides/hybrid-apps#switching_to_vr_mode_at_runtime) developer documentation.
